ai/src/data.py

```diff
@@ -238,3 +246,13 @@ def reward(
             raise RewardInvalidException(repr(ex)) from ex
 
         return loc["reward"]
+
+    def external_reward(self, action_name: str, prev_state_dict: dict, prev_state_interpretations,
```

Comment:
I would have get the method (`getattr`) directly in the initialization and put this function content (4 lines of code) directly in the `reward` function (avoiding passing all the parameters).